### ISSUE TITLE
Avoid corev1.ObjectReference

### ISSUE THREAD

**Issue by negz** [crossplaneio/crossplane-runtime#49]
### What problem are you facing?

Most inter-resource references in Crossplane are either an [`ObjectReference`](https://godoc.org/k8s.io/api/core/v1#ObjectReference) or a [`LocalObjectReference`](https://godoc.org/k8s.io/api/core/v1#ObjectReference). `ObjectReference` contains many optional fields (e.g. `UID`, `FieldPath`, etc) that Crossplane never uses. This could be misleading to users, since some fields (i.e. `Name`) are always optional, and it's not obvious that the other fields are frequently not required when reading the API documentation or CRD specs.

### How could Crossplane help solve your problem?

Crossplane could use purpose-specific references that represent a subset of `ObjectReference`.

**Comment by negz** [comment]
A related issue: https://github.com/kubernetes-sigs/service-catalog/issues/1411

**Comment by muvaf** [comment]
I agree. A PR that implements all meta methods we have that are built on this object for other reference types we have could be helpful for adoption.

**Comment by muvaf** [comment]
The refactor should include the interfaces where `corev1.ObjectReference` is accepted as parameter, too, like `SetCompositionReference(*corev1.ObjectReference)` etc.
